experiment/run_experiment.py

```diff
@@ -410,59 +407,42 @@ def start(self):
+            gcloud.create_instance(self.instance_name,
+                                   gcloud.InstanceType.DISPATCHER,
+                                   self.config,
+                                   startup_script=startup_script.name)
+
+    def _render_startup_script(self):
+        template = JINJA_ENV.get_template(
+            'dispatcher-startup-script-template.sh')
         cloud_sql_instance_connection_name = (
             self.config['cloud_sql_instance_connection_name'])
 
-        command = (
-            'echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope && '
-            'docker run --rm '
-            '-e INSTANCE_NAME="{instance_name}" '
-            '-e EXPERIMENT="{experiment}" '
-            '-e CLOUD_PROJECT="{cloud_project}" '
-            '-e DOCKER_REGISTRY="{docker_registry}" '
-            '-e EXPERIMENT_FILESTORE="{experiment_filestore}" '
-            '-e POSTGRES_PASSWORD="{postgres_password}" '
-            '-e CLOUD_SQL_INSTANCE_CONNECTION_NAME='
-            '"{cloud_sql_instance_connection_name}" '
-            '--cap-add=SYS_PTRACE --cap-add=SYS_NICE '
-            '-v /var/run/docker.sock:/var/run/docker.sock '
-            '--name=dispatcher-container '
-            '{docker_registry}/dispatcher-image '
-            '/work/startup-dispatcher.sh'
-        ).format(
-            instance_name=self.instance_name,
-            postgres_password=os.environ['POSTGRES_PASSWORD'],
-            experiment=self.config['experiment'],
+        kwargs = {
+            'instance_name': self.instance_name,
+            'postgres_password': os.environ['POSTGRES_PASSWORD'],
+            'experiment': self.config['experiment'],
             # TODO(metzman): Create a function that sets env vars based on
             # the contents of a dictionary, and use it instead of hardcoding
             # the configs we use.
-            cloud_project=self.config['cloud_project'],
-            experiment_filestore=self.config['experiment_filestore'],
-            cloud_sql_instance_connection_name=(
-                cloud_sql_instance_connection_name),
-            docker_registry=docker_registry,
-        )
-        return gcloud.ssh(self.instance_name,
-                          command=command,
-                          zone=self.config['cloud_compute_zone'])
+            'cloud_project': self.config['cloud_project'],
+            'experiment_filestore': self.config['experiment_filestore'],
+            'cloud_sql_instance_connection_name':
+                (cloud_sql_instance_connection_name),
+            'docker_registry': self.config['docker_registry'],
+        }
+        return template.render(**kwargs)
+
+    def write_startup_script(self, startup_script_file):
+        """Get the startup script to start the experiment on the dispatcher."""
+        startup_script = self._render_startup_script()
+        startup_script_file.write(startup_script)
+        startup_script_file.flush()
```

Comment:
`close()` instead of `flush()`?

Reply:
Flush is needed. Closing the file deletes it.